src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -84,23 +64,78 @@ private[concurrent] object Promise {
     case t                                         => Failure(t)
   }
 
-   /**
-    * Latch used to implement waiting on a DefaultPromise's result.
-    *
-    * Inspired by: http://gee.cs.oswego.edu/cgi-bin/viewcvs.cgi/jsr166/src/main/java/util/concurrent/locks/AbstractQueuedSynchronizer.java
-    * Written by Doug Lea with assistance from members of JCP JSR-166
-    * Expert Group and released to the public domain, as explained at
-    * http://creativecommons.org/publicdomain/zero/1.0/
-    */
-    private final class CompletionLatch[T] extends AbstractQueuedSynchronizer with (Try[T] => Unit) {
-      override protected def tryAcquireShared(ignored: Int): Int = if (getState != 0) 1 else -1
-      override protected def tryReleaseShared(ignore: Int): Boolean = {
-        setState(1)
-        true
-      }
-      override def apply(ignored: Try[T]): Unit = releaseShared(1)
+  sealed trait Callbacks[+T] {
```

Comment:
IIRC making this a `sealed abstract class` instead of a `trait` should have slight perf benefits in most JIT tiers.
(source: http://shipilev.net/blog/2015/black-magic-method-dispatch/)